### PR TITLE
Adjust used Docker container to run integration tests [MAILPOET-4251]

### PR DESCRIPTION
**Pull request by samnajian**
### This PR:
Uses the setup and the docker image used to run acceptance tests to run integration tests for single and multi-site installs as well. In doing so the Robofile is adjusted, some failing tests have been fixed, new docker images have been generated and used, the docker-compose config is adjusted and CircleCI config has been update.

### How to test:
- Make sure the tests pass for on CI for single and multi-site
- Verify the Xdebug debugging experience is acceptable by following the setup description added to the README.md file

[MAILPOET-4251]

[MAILPOET-4251]: https://mailpoet.atlassian.net/browse/MAILPOET-4251?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by costasovo — CHANGES_REQUESTED** [review]
Great job! Tests run nicely and I also tested the Xdebug support.
This is a great change! Thank you for looking into it.

I noticed some small details, that may need some changes. See comments.

I have also a couple of notes/questions:

1) I wanted to add that with this change we can remove the [test_wordpress service](https://github.com/mailpoet/mailpoet/blob/master/docker-compose.yml#L58-L81) from the main docker-compose.yml. But then I realized that we can't because we still need it for the [premium plugin's tests](https://github.com/mailpoet/mailpoet-premium/blob/master/RoboFile.php#L188-L209). So it seems we will need to do the change also for the premium plugin. I suppose this should be done as a follow-up ticket. Could you please prepare one? Thinking about it in the future we may use one docker-compose configuration for the free and premium. What do you think?

2) On the current master we are able to run the integration tests from the root of the repository using `./do --test test:integration`. This is not working for the free plugin after the change. It tries to run the command on the `test_wordpress` container and ends with `docker-compose: not found` error. It works for premium though. I don't know if any devs use the approach. Maybe we could add a condition for this case to the [do script in root](https://github.com/mailpoet/mailpoet/blob/master/do).

3) I noticed that when we run the integration tests we also start chrome and test site container, which are not needed. But this is just a detail. I think the most negative effect in has on the Circle CI where [we now download these containers](https://app.circleci.com/pipelines/github/mailpoet/mailpoet/9854/workflows/c1562efe-a472-4b97-9076-c8e5f3dccd3d/jobs/178231/parallel-runs/0/steps/0-102). I think this can be fixed in a follow-up task as well.

**Comment by costasovo** [comment]
> I noticed that when we run the integration tests we also start chrome and test site container, which are not needed. But this is just a detail. I think the most negative effect in has on the Circle CI where [we now download these containers](https://app.circleci.com/pipelines/github/mailpoet/mailpoet/9854/workflows/c1562efe-a472-4b97-9076-c8e5f3dccd3d/jobs/178231/parallel-runs/0/steps/0-102). I think this can be fixed in a follow-up task as well.

I had a brief discussion with @JanJakes (wanted to ask him for second review) and we came up with an idea of defining a new service in tests `docker-compose.yml`. Now we have [codeception service](https://github.com/mailpoet/mailpoet/blob/194aa74a56917765602993b4d59a751c6f38f1bc/mailpoet/tests/docker/docker-compose.yml#L4-L24) for both integration test and also for acceptance tests. If we define two services we could define specific env variables (instead of passing TEST_TYPE from the outside) within these services and at the same time, we could define different `depends_on` for each and start only the necessary containers. What do you think?

**Comment by samnajian** [comment]
> 1. I wanted to add that with this change we can remove the [test_wordpress service](https://github.com/mailpoet/mailpoet/blob/master/docker-compose.yml#L58-L81) from the main docker-compose.yml. But then I realized that we can't because we still need it for the [premium plugin's tests](https://github.com/mailpoet/mailpoet-premium/blob/master/RoboFile.php#L188-L209). So it seems we will need to do the change also for the premium plugin. I suppose this should be done as a follow-up ticket. Could you please prepare one? Thinking about it in the future we may use one docker-compose configuration for the free and premium. What do you think?

A unified docker-compose file makes a lot of sense, and I'll create tickets for follow ups we need to work on


> 2\. On the current master we are able to run the integration tests from the root of the repository using `./do --test test:integration`. This is not working for the free plugin after the change. It tries to run the command on the `test_wordpress` container and ends with `docker-compose: not found` error. It works for premium though. I don't know if any devs use the approach. Maybe we could add a condition for this case to the [do script in root](https://github.com/mailpoet/mailpoet/blob/master/do).

will take a look into it, we can probably make the do command in the root fallback to the commands in the `mailpoet`


> 3\. I noticed that when we run the integration tests we also start chrome and test site container, which are not needed. But this is just a detail. I think the most negative effect in has on the Circle CI where [we now download these containers](https://app.circleci.com/pipelines/github/mailpoet/mailpoet/9854/workflows/c1562efe-a472-4b97-9076-c8e5f3dccd3d/jobs/178231/parallel-runs/0/steps/0-102). I think this can be fixed in a follow-up task as well.

Actually I have ideas to mitigate this, but was more inclined in keeping it in a follow up ticket

File: mailpoet/.env.sample
Comment (samnajian) [reply mailpoet/.env.sample:7]: I'm wondering what you get for `WP_ROOT`  in `tests/integration/_bootstrap.php` ? it doesn't work for me if the value is not changed to `/wp-core`, maybe it's cached in a container for you?
I can introduce new variables as well to avoid conflict

File: mailpoet/tests/docker/codeception/docker-entrypoint.sh
Comment (samnajian) [reply mailpoet/tests/docker/codeception/docker-entrypoint.sh:36]: I feel like this approach keeps it more verbose and thus easier to spot later on. With a backup we may also run into file system issues we sometimes experience with Docker.

File: mailpoet/tests/docker/docker-compose.yml
Comment (samnajian) [reply mailpoet/tests/docker/docker-compose.yml:28]: Acceptance tests read saved tests from `http://mailhog:8025` see https://github.com/mailpoet/mailpoet/blob/master/mailpoet/tests/_support/AcceptanceTester.php#L46

File: .circleci/config.yml
Comment (samnajian) [reply .circleci/config.yml:445]: nice point

**Comment by samnajian** [comment]
> I had a brief discussion with @JanJakes (wanted to ask him for second review) and we came up with an idea of defining a new service in tests `docker-compose.yml`. Now we have [codeception service](https://github.com/mailpoet/mailpoet/blob/194aa74a56917765602993b4d59a751c6f38f1bc/mailpoet/tests/docker/docker-compose.yml#L4-L24) for both integration test and also for acceptance tests. If we define two services we could define specific env variables (instead of passing TEST_TYPE from the outside) within these services and at the same time, we could define different `depends_on` for each and start only the necessary containers. What do you think?

We probably can leverage [docker compose profiles](https://docs.docker.com/compose/profiles/) as well to achieve the same thing without adding lots of redundancies

File: mailpoet/.env.sample
Comment (costasovo) [reply mailpoet/.env.sample:7]: I added a breakpoint into `_bootstrap.php` and I'm getting `/wp-core`. It seems to be because it is defined in [docker-compose.yml](https://github.com/mailpoet/mailpoet/blob/984fa678bc75d0e8e1b97e8232a3604882140bd8/mailpoet/tests/docker/docker-compose.yml#L19)
When I remove the definition of WP_ROOT from the docker-compose.yml I start getting the value from my .env

File: mailpoet/tests/docker/codeception/docker-entrypoint.sh
Comment (costasovo) [reply mailpoet/tests/docker/codeception/docker-entrypoint.sh:36]: Ok, let's keep it this way. We can change it in the future if we find we need to touch the list often.

**Comment by costasovo** [comment]
> We probably can leverage [docker compose profiles](https://docs.docker.com/compose/profiles/) as well to achieve the same thing without adding lots of redundancies

👍  TIL. The profiles look like the right tool for this.

**Comment by JanJakes** [comment]
>> We probably can leverage [docker compose profiles](https://docs.docker.com/compose/profiles/) as well to achieve the same thing without adding lots of redundancies

> 👍 TIL. The profiles look like the right tool for this.

@samnajian @costasovo Actually, I think in this case it may be better to have this single service defined twice in the `docker-compose.yml` to also avoid having things like `TEST_TYPE` passed in the command itself and being nicely explicit about some configuration differences between those two. Dependencies will also be better visible than with profiles. What do you think?

File: mailpoet/tests/docker/codeception/docker-entrypoint.sh
Comment (JanJakes) [reply mailpoet/tests/docker/codeception/docker-entrypoint.sh:36]: Here, I think writability shouldn't be an issue as `wp config delete` needs to write as well. Seeing this is the entrypoint, I'm not sure if "backup" file makes sense, could we instead generate a "fresh" WP config in the entrypoint?

Comment (samnajian) [reply mailpoet/tests/docker/codeception/docker-entrypoint.sh:36]: @JanJakes creating a fresh config file is a good idea as well, will take a look into it.

**Comment by samnajian** [comment]
> > > We probably can leverage [docker compose profiles](https://docs.docker.com/compose/profiles/) as well to achieve the same thing without adding lots of redundancies
> 
> > 👍 TIL. The profiles look like the right tool for this.
> 
> @samnajian @costasovo Actually, I think in this case it may be better to have this single service defined twice in the `docker-compose.yml` to also avoid having things like `TEST_TYPE` passed in the command itself and being explicit about configuration differences between those two. Dependencies will also be better visible than with profiles. What do you think?

@JanJakes I see your point, but at the same time we'll have two separate entry files which is not a bad thing per se, will give it a try

**Comment by JanJakes** [comment]
> we'll have two separate entry files

@samnajian Now I'm a bit confused, I just thought that instead of one service such as [this](https://github.com/mailpoet/mailpoet/blob/f15a7634e5d55d72ccf0668855ba01b885db23db/mailpoet/tests/docker/docker-compose.yml#L4-L21), there would be two in the same `docker-compose.yml` file. But I don't know the details so maybe I'm missing something. Is this what you had in mind or some other solution?

**Comment by samnajian** [comment]
> > we'll have two separate entry files
> 
> @samnajian Now I'm a bit confused, I just thought that instead of one service such as [this](https://github.com/mailpoet/mailpoet/blob/f15a7634e5d55d72ccf0668855ba01b885db23db/mailpoet/tests/docker/docker-compose.yml#L4-L21), there would be two in the same `docker-compose.yml` file. But I don't know the details so maybe I'm missing something. Is this what you had in mind or some other solution?

1) I was of the idea of making it work with one `codeception` service and `profiles`.
2) Then you suggested two seperate `codeception` services ( with different names of course ) and removing `TEST_TYPE` which would keeping one single entry points not possible nor clean.

I'm finding that dependency resolution and `profiles` don't work well together and don't serve our purpose that's why I'm going to give the second option a try

**Comment by JanJakes** [comment]
@samnajian Thanks, I get it now. You didn't mean creating extra entry "files" rather than having different entry points in terms of executing a different service. I guess if you don't run into any issues there, it's a viable option.

File: mailpoet/tests/docker/codeception/docker-entrypoint.sh
Comment (samnajian) [reply mailpoet/tests/docker/codeception/docker-entrypoint.sh:36]: @JanJakes https://developer.wordpress.org/cli/commands/config/create/ didn't work stable, it was breaking the tests from time to time with failing respect --skip-check for db, so I'm leaving it like this for now

Comment (JanJakes) [reply mailpoet/tests/docker/codeception/docker-entrypoint.sh:36]: OK, we can always try to resolve that later.

**Comment by samnajian** [comment]
> * I wanted to add that with this change we can remove the [test_wordpress service](https://github.com/mailpoet/mailpoet/blob/master/docker-compose.yml#L58-L81) from the main docker-compose.yml. But then I realized that we can't because we still need it for the [premium plugin's tests](https://github.com/mailpoet/mailpoet-premium/blob/master/RoboFile.php#L188-L209). So it seems we will need to do the change also for the premium plugin. I suppose this should be done as a follow-up ticket. Could you please prepare one? Thinking about it in the future we may use one docker-compose configuration for the free and premium. What do you think?
> *

[Ticket created for premium plugin](https://mailpoet.atlassian.net/browse/MAILPOET-4336)

**Review by costasovo — DISMISSED** [review]
Thanks! Great job!

I was thinking that maybe we could split the docker-entrypoint.sh file into two because a big part is conditional and needed only for the acceptance test, but on the other hand a lot is shared. So I gues it is ok as is and we can always split it if needed.

I just want to ask @JanJakes for one final review 🙏

**Review by JanJakes — COMMENTED** [review]
LGTM, great job!

I have just one issue:

Screen Shot 2022-05-13 at 17 39 08

@samnajian Does it occur also for you if you remove the proxies in the generated folder? Running `./do doctrine:generate-cache` fixes that. In dev mode they don't need to be generated (they're autogenerated), not sure now what we had set up in the tests, seems like it doesn't autogenerate.

**Comment by samnajian** [comment]
> @samnajian Does it occur also for you if you remove the proxies in the generated folder? Running `./do doctrine:generate-cache` fixes that. In dev mode they don't need to be generated (they're autogenerated), not sure now what we had set up in the tests, seems like it doesn't autogenerate.

@JanJakes  yes I had the same issue, not sure if it's an issue and needs checking?

File: mailpoet/tests/docker/docker-compose.yml
Comment (JanJakes) [inline mailpoet/tests/docker/docker-compose.yml:37]: I just noticed that both integration and acceptance tests share the same `wp-core` volume. This probably means that could affect each other (installed plugins, changes in `wp-config.php`, etc.). Could we make those two different volumes? (Such as `wp-core-acceptance` vs. `wp-core-integration`?)

**Review by JanJakes — COMMENTED** [review]
@samnajian What if we just add `doctrineGenerateCache` to `runTestsInContainer` for now? That could be a reasonably quick fix for both integration and acceptance tests.

File: mailpoet/tests/docker/docker-compose.yml
Comment (samnajian) [reply mailpoet/tests/docker/docker-compose.yml:37]: @JanJakes Yes that's true, it's shared, thought we'd better off using less resources and the `wp CLI` scripts should be able to protect us from running into issue because of sharing this volume, I'm also not against your idea and we can always do this down the road or in the override

**Review by JanJakes — DISMISSED** [review]
Great stuff! 👏

**Review by veljkho — APPROVED** [review]

**Pull request merged**
